storage/raft_log_queue.go

```diff
@@ -114,28 +114,31 @@ func getTruncatableIndexes(r *Replica) (uint64, uint64, error) {
 // is down long enough, sending a snapshot is more efficient and we should
 // truncate the log to the next behind node or the quorum committed index. We
 // currently truncate when the raft log size is bigger than the range
-// size. When there are no nodes behind, or we can't catch any of them up via
-// the raft log (due to a previous truncation) this returns the quorum
-// committed index.
+// size.
+//
+// Note that when a node is behind we continue to let the raft log build up
+// instead of truncating to the commit index. Consider what would happen if we
+// truncated to the commit index whenever a node is behind and thus needs to be
+// caught up via a snapshot. While we're generating the snapshot, sending it to
+// the behind node and waiting for it to be applied we would continue to
+// truncate the log. If the snapshot generation and application takes too long
+// the behind node will be caught up to a point behind the current first index
+// and thus require another snapshot, likely entering a never ending loop of
+// snapshots. See #8629.
 func computeTruncatableIndex(
 	raftStatus *raft.Status,
 	raftLogSize, targetSize int64,
 	firstIndex uint64,
 ) uint64 {
 	truncatableIndex := raftStatus.Commit
 	behindIndexes := getBehindIndexes(raftStatus)
-	if raftLogSize <= targetSize {
+	if raftLogSize <= targetSize && len(behindIndexes) > 0 {
```

Comment:
By the way, why in the heck are we computing a unique array of behind indexes in `getBehindIndexes`? That's just silly. Could you remove that extra complication?